### ISSUE TITLE
CSV format support + missing field values in `%( )l` comma separated list of `key:value` pairs

### ISSUE THREAD

**Issue by chrizilla** [yt-dlp/yt-dlp#6855]
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm asking a question and **not** reporting a bug or requesting a feature
- [X] I've looked through the [README](https://github.com/yt-dlp/yt-dlp#readme)
- [X] I've verified that I'm running yt-dlp version **2023.03.04** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar questions **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Please make sure the question is worded well enough to be understood

I can output `requested_format` as json `key:value` pairs but when I try to output it as comma-separated-values with `%(requested_formats)l` or `%(requested_formats)#l` I am missing the values (see log below). How can I get the values? I'd like to output csv files.

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```json
yt-dlp ZYlALC1Pmv4 --skip-download --no-config --print %(requested_formats.1)#j
{
    "asr": 48000,
    "filesize": 41488798,
    "format_id": "251",
    "format_note": "medium",
    "source_preference": -1,
    "fps": null,
    "audio_channels": 2,
    "height": null,
    "quality": 3.0,
    "has_drm": false,
    "tbr": 112.176,
    "url": "https://rr3---sn-h0jeenle.googlevideo.com/videoplayback?......",
    "width": null,
    "language": null,
    "language_preference": -1,
    "preference": null,
    "ext": "webm",
    "vcodec": "none",
    "acodec": "opus",
    "dynamic_range": null,
    "container": "webm_dash",
    "downloader_options": {
        "http_chunk_size": 10485760
    },
    "protocol": "https",
    "resolution": "audio only",
    "aspect_ratio": null,
    "http_headers": {
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36",
        "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
        "Accept-Language": "en-us,en;q=0.5",
        "Sec-Fetch-Mode": "navigate"
    },
    "audio_ext": "webm",
    "video_ext": "none",
    "abr": 112.176,
    "format": "251 - audio only (medium)"
}
```

```sh
yt-dlp ZYlALC1Pmv4 --skip-download --no-config --print %(requested_formats.1)#l
asr
filesize
format_id
format_note
source_preference
fps
audio_channels
height
quality
has_drm
tbr
url
width
language
language_preference
preference
ext
vcodec
acodec
dynamic_range
container
downloader_options
protocol
resolution
aspect_ratio
http_headers
audio_ext
video_ext
abr
format
```

**Comment by dirkf** [comment]
In the problem case `requested_formats` is a list of 2 items.

`--print '%(requested_formats.1)j'` outputs the `[1]` item as JSON, as expected.

Neither `--print '%(requested_formats.1)l'` nor `--print '%(requested_formats.1)#l'` output the `[1]` item as CSV, omitting the values in both cases.

Without the list traversal, some "CSV" output is generated. It appears to be just the JSON of one of the list elements.

You might expect that the expected output is what would be generated by 
```
yt-dlp ZYlALC1Pmv4 --skip-download --no-config --print '%(requested_formats)j' | jq '.[1]|@csv'
```
but _jq_ doesn't know how to render a JSON object as CSV. Perhaps a similar problem is being revealed in the output template processing. How is a nested JSON/dict structure supposed to be represented in CSV, anyway? Something like this?
```py
def jsonobj2csv(value):
    return '\n'.join(
        (','.join(repr(x)) for x in (value.keys(), value.values())))
```

**Comment by pukkandan** [comment]
Our code is just doing `list(dct)` like with any other iterable, which is why just the keys are being printed. This is pretty useless behavior, so I am willing to change it without being concerned about compat. But I don't know what would be an "expected" result - hence my question to OP. 


Btw, `l` is just comma/newline separated text, not csv. Escaping rules are not followed.

**Comment by dirkf** [comment]
That explains why it's not described as CSV in the doc, then!

`list(itertools.chain(*dct.items()))` seems reasonable to me (k1, v1, k2, v2, ...), but I have no standing. Nested objects as strings?

**Comment by pukkandan** [comment]
https://github.com/yt-dlp/yt-dlp/issues/6855#issuecomment-1515213514 - None of those seem like reasonable extensions

> > Escaping rules are not followed.
> 
> One solution to do away with the need for escaping would be a tab-separated list instead of a comma-separated list, because `tab` wouldn't be encountered anyway, so no need to escape, no?

Tab can appear in description, comments etc.



> > > not csv. Escaping rules are not followed.
> > 
> > 
> > That explains why it's not described as CSV in the doc, then!
> 
> Oh ok. If it's not CSV, I'll drop my request. I thought I could output .csv files. That would be nice to have.

Supporting CSV should be easy. The original requested of the feature just wanted comma separated, so making it CSV never even crossed my mind. The question of what to do with dicts still remain though

**Comment by dirkf** [comment]
If _jq_ doesn't know how to do it, maybe there isn't a good solution.

**Comment by pukkandan** [comment]
The question is whether there a **useful** solution. @chrizilla What output were you expecting to get when opening this issue? I'm not asking for an answer you come up with now[^1], but what your original intend was.


[^1]: Like https://github.com/yt-dlp/yt-dlp/issues/6855#issuecomment-1515213514, which seems to be a solution just for the sake of one and doesn't seem like a practically useful format. Feel free to correct me if I'm wrong

**Comment by pukkandan** [comment]
> ```
> asr,48000
> filesize,41488798
> format_id,251
> format_note,medium
> ```

I don't understand how you'd actually use this. Remember you aren't just printing a single format but all of them. It'll just be like:

```
format_id,video1
asr, ...
...
format_id,video2
...
```

That seemed unstructured and pointless to me. Unless I am misunderstanding?

**Comment by pukkandan** [comment]
I think I misunderstood. Let me go over the convo again

**Comment by pukkandan** [comment]
> I can output `requested_format` as json `key:value` pairs but when I try to output it as comma-separated-values with `%(requested_formats)l` or `%(requested_formats)#l` I am missing the values (see log below). How can I get the values?

This is wrong. According to your logs, `%(requested_formats)#l` already work as documented[^2]. You are actually trying to get `%(requested_formats.1)#l` to work. I had understood the issue originally, but misunderstood when I went through it again yesterday [^1]

[^2]: newline separated dicts
[^1]: Yes, I often re-read the OP. I'm answering too many issues to keep it all in my head.

In that case, your suggestion sort of makes sense. I still don't think it's a universally useful format, but since nobody has better ideas, ig why not. Let me see what can be done

**Comment by pukkandan** [comment]
> ![Clipboard02](https://user-images.githubusercontent.com/43895159/234099929-1091cced-dee8-4bf4-8d3f-c434d588c3ef.png)


This is not how CSV is generally used and is way too much of a special format for us to support

A|B|C|D
---|---|---|---
format_id|container|download_options|protocol
f1|webm_dash|{"http_chunk_size": 10485760}|https
f2|mp4|{"http_chunk_size": 10485760}|https

is the most reasonable format I can come up with, though idk if that's useful to anyone
